File: backend/go-chi-hive/internal/shared/web/auth_middleware.go
package web

import (
	"context"
	"net/http"
	"strings"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hive/internal/shared/errors"
)

type contextKey string

const userIDKey contextKey = "user_id"

type TokenParser interface {
	Parse(token string) (int64, error)
}

func AuthMiddleware(tokenParser TokenParser) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			authHeader := r.Header.Get("Authorization")
			if authHeader == "" {
				RespondWithError(w, r, errors.NewUnauthorizedError("authorization header is required"))
				return
			}

			parts := strings.Split(authHeader, " ")
			if len(parts) != 2 || parts[0] != "Token" {
				RespondWithError(w, r, errors.NewUnauthorizedError("invalid authorization header format. Expected 'Token <token>'"))
				return
			}

			userID, err := tokenParser.Parse(parts[1])
			if err != nil {
				RespondWithError(w, r, errors.NewUnauthorizedError("invalid or expired token"))
				return
			}

			ctx := context.WithValue(r.Context(), userIDKey, userID)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

func OptionalAuthMiddleware(tokenParser TokenParser) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			authHeader := r.Header.Get("Authorization")
			if authHeader == "" {
				next.ServeHTTP(w, r)
				return
			}

			parts := strings.Split(authHeader, " ")
			if len(parts) != 2 || parts[0] != "Token" {
				next.ServeHTTP(w, r)
				return
			}

			token := parts[1]
			userID, err := tokenParser.Parse(token)
			if err != nil {
				next.ServeHTTP(w, r)
				return
			}

			ctx := context.WithValue(r.Context(), userIDKey, userID)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

func GetUserIDFromContext(ctx context.Context) (int64, bool) {
	userID, ok := ctx.Value(userIDKey).(int64)
	return userID, ok
}

func WithUserID(ctx context.Context, userID int64) context.Context {
	return context.WithValue(ctx, userIDKey, userID)
}

File: backend/typescript-node-express-hive/src/shared/web/auth-middleware.ts
import type { NextFunction, Request, Response } from "express";
import { JwtTokenGenerator } from "../security/token.js";
import { newUnauthorizedError } from "../errors/app-error.js";

export interface AuthenticatedRequest extends Request {
    userId?: number;
}

export function authMiddleware(tokenGenerator: JwtTokenGenerator, required: boolean = true) {
    return async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
        const authHeader = req.headers.authorization;
        if (!authHeader) {
            if (required) {
                return next(newUnauthorizedError("Authorization header is required"));
            }
            return next();
        }

        const parts = authHeader.split(" ");
        if (parts.length !== 2 || parts[0] !== "Token") {
            if (required) {
                return next(newUnauthorizedError("Invalid authorization header format"));
            }
            return next();
        }

        const token = parts[1];
        if (!token) {
            if (required) {
                return next(newUnauthorizedError("Token is required"));
            }
            return next();
        }

        try {
            req.userId = await tokenGenerator.verify(token);
            next();
        } catch (err) {
            if (required) {
                return next(newUnauthorizedError("Invalid or expired token"));
            }
            next();
        }
    };
}
